### PR TITLE
Feature add registration

### PR DESCRIPTION
**Pull request by gouthamr321**
This is an early R&D version of a newer registration method. I have implemented the existing tf registration method for this dataset.

### CODE REVIEW

**Comment by gouthamr321** [comment]
Major changes:
- Include ORB based registration
- Jinja templates for parallelizing steps of this workflow that take the longest

**Comment by gouthamr321** [comment]
This should be ready to be reviewed before we merge into main. I've updated the readme to reflect the changes made in this PR... but basically the major changes are the following:

1) ORB based feature registration
2) Jinja templates for the compute alignment parameters and round alignment steps for whichever method the user decides to use for the alignment
3) script to generate gifs to evaluate the alignments

File: multiplex_immuno_processing/compute_alignment.py
Comment (cfrick13) [inline multiplex_immuno_processing/compute_alignment.py]: Lets try to use aicsimageio for all file writing tasks

File: multiplex_immuno_processing/compute_orb_based_registration_paramaters.py
Comment (cfrick13) [inline multiplex_immuno_processing/compute_orb_based_registration_paramaters.py]: why is this step still in the code?

File: multiplex_immuno_processing/compute_orb_based_registration_paramaters.py
Comment (cfrick13) [inline multiplex_immuno_processing/compute_orb_based_registration_paramaters.py]: need to change 100 (here and elsewhere) to be encoded by a variable (maybe called `align_crop_val`?). Otherwise one of these 100s will get lost somewhere almost certainly.

Comment (cfrick13) [reply multiplex_immuno_processing/compute_orb_based_registration_paramaters.py]: Also, why doesn't the ORB method compute the offset in Z?

File: multiplex_immuno_processing/compute_orb_based_registration_paramaters.py
Comment (cfrick13) [inline multiplex_immuno_processing/compute_orb_based_registration_paramaters.py]: one column here should be the method used to compute the alignment . E.g. `Method` : `Cross corr` or `Method` : `ORB`

File: multiplex_immuno_processing/execute_single_position_alignment.py
Comment (cfrick13) [inline multiplex_immuno_processing/execute_single_position_alignment.py]: no need to name this file `single_position`. It can just take an argument to do single position and keep the same name as before.

File: multiplex_immuno_processing/execute_single_position_alignment_ORB.py
Comment (cfrick13) [inline multiplex_immuno_processing/execute_single_position_alignment_ORB.py]: this should take an argument to use ORB alignment paraemters. Not be its own code.

File: multiplex_immuno_processing/generate_alignment_parameters_for_single_position.py
Comment (cfrick13) [inline multiplex_immuno_processing/generate_alignment_parameters_for_single_position.py]: should take an argument to run on a single possition. shoudl not be its own file.

File: multiplex_immuno_processing/generate_contact_sheet_gif.py
Comment (cfrick13) [inline multiplex_immuno_processing/generate_contact_sheet_gif.py:6]: aicsimageio has a timeseries writer that can write gifs. I'd recommend using that.

File: multiplex_immuno_processing/run_alignment.j2
Comment (cfrick13) [inline multiplex_immuno_processing/run_alignment.j2]: these jinja templates won't work for anyone else because they are specified for your working directory. You should pass in an argument from the render_dict that specifies where the code is being run from . E.g. `srun conda/env {{ code_dir }}python_filename.py --etc etc`

File: multiplex_immuno_processing/run_alignment_ORB.j2
Comment (cfrick13) [inline multiplex_immuno_processing/run_alignment_ORB.j2]: again. this won't work for anyone else.

**Review by cfrick13 — CHANGES_REQUESTED** [review]

File: multiplex_immuno_processing/compute_orb_based_registration_paramaters.py
Comment (gouthamr321) [reply multiplex_immuno_processing/compute_orb_based_registration_paramaters.py]: It does calculate the offset in Z, but the z offset is very poor. After some trial and error, I decided to use the cross correlation method's Z alignment and keep the XY alignment from the ORB method--- This produced the best results

File: multiplex_immuno_processing/compute_alignment.py
Comment (gouthamr321) [reply multiplex_immuno_processing/compute_alignment.py]: We don't actually need this script in the workflow with the new jinja templates. Will remove this

File: multiplex_immuno_processing/execute_single_position_alignment.py
Comment (gouthamr321) [reply multiplex_immuno_processing/execute_single_position_alignment.py]: This is actually the script that the jinja tempelate calls and its by default set to only do alignment for a single position. But yeah, the naming convention is weird--- will change that

File: multiplex_immuno_processing/compute_orb_based_registration_paramaters.py
Comment (gouthamr321) [reply multiplex_immuno_processing/compute_orb_based_registration_paramaters.py]: See below thread

File: multiplex_immuno_processing/compute_orb_based_registration_paramaters.py
Comment (cfrick13) [reply multiplex_immuno_processing/compute_orb_based_registration_paramaters.py]: got it. that works for now.

File: multiplex_immuno_processing/execute_single_position_alignment.py
Comment (cfrick13) [reply multiplex_immuno_processing/execute_single_position_alignment.py]: sounds good

**Comment by cfrick13** [comment]
I think we can approve this once we get all the of redundant files removed

**Comment by gouthamr321** [comment]
I've made most of the changes already.... currently just testing it out on a plate from scratch to make sure I didn't break anything. Holding off on committing until then, but if everything looks good, hopefully we can get this merged by the end of the day

File: multiplex_immuno_processing/compute_orb_based_registration_paramaters.py
Comment (gouthamr321) [reply multiplex_immuno_processing/compute_orb_based_registration_paramaters.py]: This is resolved in the new set of commits

File: multiplex_immuno_processing/execute_single_position_alignment_ORB.py
Comment (gouthamr321) [reply multiplex_immuno_processing/execute_single_position_alignment_ORB.py]: resolved in new set of commits

File: multiplex_immuno_processing/generate_alignment_parameters_for_single_position.py
Comment (gouthamr321) [reply multiplex_immuno_processing/generate_alignment_parameters_for_single_position.py]: resolved in new set of commits

File: multiplex_immuno_processing/generate_contact_sheet_gif.py
Comment (gouthamr321) [reply multiplex_immuno_processing/generate_contact_sheet_gif.py:6]: I didn't include a change for this yet, but we can do this in a separate branch

File: multiplex_immuno_processing/run_alignment.j2
Comment (gouthamr321) [reply multiplex_immuno_processing/run_alignment.j2]: resolved

File: multiplex_immuno_processing/run_alignment_ORB.j2
Comment (gouthamr321) [reply multiplex_immuno_processing/run_alignment_ORB.j2]: resolved

File: multiplex_immuno_processing/execute_single_position_alignment.py
Comment (gouthamr321) [reply multiplex_immuno_processing/execute_single_position_alignment.py]: I decided to keep the single position naming convention here because this is only doing it on a single position w/ the jinja tempelate, but we can change this later

**Comment by cfrick13** [comment]
Lots of files in here now have names that are not consistent with main

`exectue_round_alignment.py` was deleted and now we have `execute_single_position_alignment.py`. 

I think it makes more sense to keep the names consistent with main. Thoughts?

**Comment by gouthamr321** [comment]
It's different now because these files are performing separate operations...... but what we could do is keep the original files in main and not override them?

**Comment by gouthamr321** [comment]
exectue_round_alignment.py is spelled wrong in the main branch---- so this is the only thing i'll change naming wise, but everything else should still be the same

**Comment by gouthamr321** [comment]
Ok the changes we just discussed should be all good now @cfrick13. Let me know your thoughts

File: multiplex_immuno_processing/execute_round_alignment.py
Comment (cfrick13) [inline multiplex_immuno_processing/execute_round_alignment.py:77]: this only aligns one position. I thought we decided to make it align all positions unless a single position was specified. So the -position argument could accept 1. a single position, 2. a list of positions or 3. no argument (in which case all positions would be aligned)

File: multiplex_immuno_processing/generate_alignment_paramters.py
Comment (cfrick13) [inline multiplex_immuno_processing/generate_alignment_paramters.py:35]: same comment as other file. This should accept a 1. a single position, 2 a list of positions or 3. no argument (in which case all positions will be processed).

**Comment by cfrick13** [comment]
I think this is good to merge. We can make slight adjustments (conversations above)  later.

File: multiplex_immuno_processing/generate_alignment_paramters.py
Comment (cfrick13) [inline multiplex_immuno_processing/generate_alignment_paramters.py:220]: there should be a column named "method" that tells which method was used rather than changing the name of `alignment_offsets_xyz`. (i should note that one needed change in a future update is to rename `alignment_offsets_xyz` to `alignment_offsets_zyx` since that is the actual order of the offsets).

File: multiplex_immuno_processing/generate_alignment_paramters.py
Comment (cfrick13) [inline multiplex_immuno_processing/generate_alignment_paramters.py:224]: we should not overwrite the file each time we run alignments with a different method. We should save the alignment files with different names to indicate a different method was used (or save them in a different directory that indicates which method was used). It's expensive computationally to run the alignments so we should make sure we can run a new method wihtout losing results from previous methods.

**Review by cfrick13 — APPROVED** [review]

File: multiplex_immuno_processing/execute_round_alignment.py
Comment (gouthamr321) [reply multiplex_immuno_processing/execute_round_alignment.py:77]: Look into keeping this general enough where one can run this for all positions without jinja setup

File: multiplex_immuno_processing/generate_alignment_paramters.py
Comment (gouthamr321) [reply multiplex_immuno_processing/generate_alignment_paramters.py:35]: see above

File: multiplex_immuno_processing/generate_alignment_paramters.py
Comment (gouthamr321) [reply multiplex_immuno_processing/generate_alignment_paramters.py:224]: Will need to address this. Use method from argparse to specify a output name

File: multiplex_immuno_processing/generate_alignment_paramters.py
Comment (gouthamr321) [reply multiplex_immuno_processing/generate_alignment_paramters.py:220]: Will come back to this

**Pull request merged**
